File: OneDrive/Desktop/data science project/placement_system/modules/analytics.py
import sqlite3
import os

DB_PATH = os.path.join(os.path.dirname(os.path.dirname(__file__)), 'placement.db')

class AnalyticsEngine:
    def __init__(self):
        pass

    def _get_connection(self):
        return sqlite3.connect(DB_PATH)

    def get_summary_stats(self):
        conn = self._get_connection()
        cursor = conn.cursor()

        # Counts
        cursor.execute("SELECT COUNT(*) FROM students")
        total_students = cursor.fetchone()[0]

        cursor.execute("SELECT COUNT(*) FROM students WHERE status = 'Placed'")
        placed_students = cursor.fetchone()[0]

        cursor.execute("SELECT COUNT(*) FROM students WHERE status = 'Not Eligible'")
        not_eligible = cursor.fetchone()[0]

        cursor.execute("SELECT COUNT(*) FROM companies")
        total_companies = cursor.fetchone()[0]

        cursor.execute("SELECT COUNT(*) FROM interviews")
        total_interviews = cursor.fetchone()[0]

        # Calculate Placement Rate
        eligible_students = total_students - not_eligible
        placement_rate = 0.0
        if eligible_students > 0:
            placement_rate = round((placed_students / eligible_students) * 100, 1)

        # Average Salary
        cursor.execute("SELECT AVG(ctc_lpa) FROM placements")
        avg_ctc = cursor.fetchone()[0] or 0.0
        avg_ctc = round(avg_ctc, 2)

        conn.close()

        return {
            'total_students': total_students,
            'placed_students': placed_students,
            'not_eligible': not_eligible,
            'eligible_students': eligible_students,
            'placement_rate': placement_rate,
            'total_companies': total_companies,
            'total_interviews': total_interviews,
            'avg_ctc_lpa': avg_ctc
        }

    def get_branch_reports(self):
        conn = self._get_connection()
        cursor = conn.cursor()

        # Aggregate branch wise
        cursor.execute("""
            SELECT 
                branch,
                COUNT(*) AS total,
                SUM(CASE WHEN status = 'Placed' THEN 1 ELSE 0 END) AS placed,
                SUM(CASE WHEN status = 'Not Eligible' THEN 1 ELSE 0 END) AS not_eligible
            FROM students
            GROUP BY branch
        """)
        rows = cursor.fetchall()

        branch_stats = []
        for r in rows:
            branch = r[0]
            total = r[1]
            placed = r[2]
            not_eligible = r[3]
            eligible = total - not_eligible
            
            rate = 0.0
            if eligible > 0:
                rate = round((placed / eligible) * 100, 1)

            # Get average package for this branch
            cursor.execute("""
                SELECT AVG(p.ctc_lpa) 
                FROM placements p
                JOIN students s ON s.student_id = p.student_id
                WHERE s.branch = ?
            """, (branch,))
            avg_ctc = cursor.fetchone()[0] or 0.0
            avg_ctc = round(avg_ctc, 2)

            branch_stats.append({
                'branch': branch,
                'total_students': total,
                'placed_students': placed,
                'eligible_students': eligible,
                'placement_rate': rate,
                'avg_ctc_lpa': avg_ctc
            })

        conn.close()
        # Sort by placement rate descending
        branch_stats.sort(key=lambda x: x['placement_rate'], reverse=True)
        return branch_stats

    def get_highest_packages(self):
        conn = self._get_connection()
        cursor = conn.cursor()
        cursor.execute("""
            SELECT s.name, s.branch, c.name, p.ctc_lpa, p.job_role, p.offer_date
            FROM placements p
            JOIN students s ON s.student_id = p.student_id
            JOIN companies c ON c.company_id = p.company_id
            ORDER BY p.ctc_lpa DESC
        """)
        rows = cursor.fetchall()
        conn.close()

        records = []
        for r in rows:
            records.append({
                'student_name': r[0],
                'student_branch': r[1],
                'company_name': r[2],
                'ctc_lpa': r[3],
                'job_role': r[4],
                'offer_date': r[5]
            })
        return records

File: OneDrive/Desktop/data science project/server/routes/analytics.js
const express = require('express');
const db = require('../db');

const router = express.Router();

// GET /api/analytics/dashboard
router.get('/dashboard', (req, res) => {
  try {
    // 1. Basic Counts
    const studentCounts = db.prepare(`
      SELECT 
        COUNT(*) AS total,
        SUM(CASE WHEN status = 'placed' THEN 1 ELSE 0 END) AS placed,
        SUM(CASE WHEN status = 'active' THEN 1 ELSE 0 END) AS active,
        SUM(CASE WHEN status = 'not_eligible' THEN 1 ELSE 0 END) AS not_eligible
      FROM students
    `).get();

    const companyCount = db.prepare('SELECT COUNT(*) AS count FROM companies').get().count;
    const activeJobsCount = db.prepare("SELECT COUNT(*) AS count FROM job_postings WHERE status = 'open'").get().count;
    const totalAppsCount = db.prepare('SELECT COUNT(*) AS count FROM applications').get().count;

    // 2. Salary Package Metrics (for placed students)
    const salaryMetrics = db.prepare(`
      SELECT 
        AVG(ctc) AS avg_ctc,
        MAX(ctc) AS max_ctc,
        MIN(ctc) AS min_ctc
      FROM placements
    `).get();

    // 3. Department-wise Placement Rates
    // We group by department and count total eligible students (placed + active) vs placed
    const deptStats = db.prepare(`
      SELECT 
        department,
        COUNT(*) AS total_students,
        SUM(CASE WHEN status = 'placed' THEN 1 ELSE 0 END) AS placed_students,
        ROUND(
          (SUM(CASE WHEN status = 'placed' THEN 1.0 ELSE 0.0 END) / 
          SUM(CASE WHEN status IN ('active', 'placed') THEN 1.0 ELSE 0.0 END)) * 100, 
          1
        ) AS placement_rate
      FROM students
      GROUP BY department
    `).all();

    // 4. Salary Distribution Groups
    const salaryDistribution = db.prepare(`
      SELECT 
        SUM(CASE WHEN ctc < 50000 THEN 1 ELSE 0 END) AS tier_1,       -- Under 50k
        SUM(CASE WHEN ctc >= 50000 AND ctc < 90000 THEN 1 ELSE 0 END) AS tier_2, -- 50k - 90k
        SUM(CASE WHEN ctc >= 90000 AND ctc < 130000 THEN 1 ELSE 0 END) AS tier_3, -- 90k - 130k
        SUM(CASE WHEN ctc >= 130000 THEN 1 ELSE 0 END) AS tier_4      -- 130k+
      FROM placements
    `).get();

    // 5. Recent placements feed
    const recentPlacements = db.prepare(`
      SELECT p.id, s.name AS student_name, s.department, c.name AS company_name, p.ctc, p.offer_date
      FROM placements p
      JOIN students s ON s.id = p.student_id
      JOIN companies c ON c.id = p.company_id
      ORDER BY p.offer_date DESC
      LIMIT 5
    `).all();

    // 6. Upcoming Interviews feed
    const upcomingInterviews = db.prepare(`
      SELECT i.id, s.name AS student_name, c.name AS company_name, jp.title AS job_title, i.round, i.scheduled_at, i.mode
      FROM interviews i
      JOIN applications a ON a.id = i.application_id
      JOIN students s ON s.id = a.student_id
      JOIN job_postings jp ON jp.id = a.job_posting_id
      JOIN companies c ON c.id = jp.company_id
      WHERE i.result IS NULL OR i.result = 'pending'
      ORDER BY i.scheduled_at ASC
      LIMIT 5
    `).all();

    // Calculate overall placement rate
    const eligibleStudentsCount = (studentCounts.total || 0) - (studentCounts.not_eligible || 0);
    const placementRate = eligibleStudentsCount > 0 
      ? Math.round(((studentCounts.placed || 0) / eligibleStudentsCount) * 100 * 10) / 10 
      : 0;

    res.json({
      counts: {
        total_students: studentCounts.total || 0,
        placed_students: studentCounts.placed || 0,
        active_students: studentCounts.active || 0,
        not_eligible_students: studentCounts.not_eligible || 0,
        eligible_students: eligibleStudentsCount,
        companies: companyCount,
        active_jobs: activeJobsCount,
        total_applications: totalAppsCount,
        placement_rate: placementRate
      },
      salary: {
        avg_ctc: Math.round(salaryMetrics.avg_ctc || 0),
        max_ctc: salaryMetrics.max_ctc || 0,
        min_ctc: salaryMetrics.min_ctc || 0
      },
      dept_stats: deptStats,
      salary_distribution: {
        under_50k: salaryDistribution.tier_1 || 0,
        tier_50k_90k: salaryDistribution.tier_2 || 0,
        tier_90k_130k: salaryDistribution.tier_3 || 0,
        above_130k: salaryDistribution.tier_4 || 0
      },
      recent_placements: recentPlacements,
      upcoming_interviews: upcomingInterviews
    });
  } catch (err) {
    console.error('Analytics Error:', err);
    res.status(500).json({ error: 'Failed to compute analytics' });
  }
});

module.exports = router;
